diff --git a/arch/arm/boot/dts/rk3288-miniarm.dts b/arch/arm/boot/dts/rk3288-miniarm.dts
index c8a5411..2f93310 100644
--- a/arch/arm/boot/dts/rk3288-miniarm.dts
+++ b/arch/arm/boot/dts/rk3288-miniarm.dts
@@ -515,11 +468,6 @@
         status = "okay";
 };
 
-&i2c5 {
-       status = "okay";
-};
-
-
 &spi2 {
         status = "okay";
         max-freq = <50000000>;
@@ -566,11 +514,6 @@
 	status = "okay";
 };
 
-&saradc {
-	vref-supply = <&vcc18_ldo1>;
-	status ="okay";
-};
-
 &sdmmc {
 	bus-width = <4>;
 	cap-mmc-highspeed;
@@ -615,10 +558,6 @@
 	status = "okay";
 };
 
-&uart4 {
-	status = "okay";
-};
-
 &usbphy {
 	status = "okay";
 };
