diff --git a/arch/arm/boot/dts/rk3288-miniarm.dts b/arch/arm/boot/dts/rk3288-miniarm.dts
index bc48294..4729b59 100644
--- a/arch/arm/boot/dts/rk3288-miniarm.dts
+++ b/arch/arm/boot/dts/rk3288-miniarm.dts
@@ -225,16 +225,6 @@
 	status = "okay";
 };
 
-&mipi_dsi {
-	status = "okay";
-	mipi_panel: mipi-panel {
-		compatible ="asus,tc358762";
-		reg = <0>;
-		status = "okay";
-	};
-};
-
-
 &gpu {
 	mali-supply = <&vdd_gpu>;
 	status = "okay";
@@ -515,25 +505,6 @@
         status = "okay";
 };
 
-&spi2 {
-        status = "okay";
-        max-freq = <50000000>;
-        spidev@0 {
-                 compatible = "rockchip,spi_tinker";
-                 reg = <0>;
-                 spi-max-frequency = <50000000>;
-                 spi-cpha = <1>;
-                 //spi-cpol = <1>;
-        };
-        spidev@1 {
-                 compatible = "rockchip,spi_tinker";
-                 reg = <1>;
-                 spi-max-frequency = <50000000>;
-                 spi-cpha = <1>;
-                 //spi-cpol = <1>;
-        };
-};
-
 &i2s {
 	#sound-dai-cells = <0>;
 	status = "okay";
