diff --git a/arch/arm/boot/dts/rk3288-miniarm.dts b/arch/arm/boot/dts/rk3288-miniarm.dts
index ddd7713..f060a8a 100644
--- a/arch/arm/boot/dts/rk3288-miniarm.dts
+++ b/arch/arm/boot/dts/rk3288-miniarm.dts
@@ -149,7 +149,7 @@
 
                 led1-led {
                         gpios=<&gpio1 25 GPIO_ACTIVE_HIGH>;
-                        linux,default-trigger="default-off";
+                        linux,default-trigger="heartbeat";
                 };
 	};
 
@@ -220,11 +220,6 @@
 	status = "ok";
 };
 
-&cif_isp0 {
-	rockchip,camera-modules-attached = <&camera1>;
-	status = "okay";
-};
-
 &gpu {
 	mali-supply = <&vdd_gpu>;
 	status = "okay";
@@ -439,96 +434,6 @@
 	};
 };
 
-&i2c2 {
-	status = "okay";
-
-	afc0:af-controller@0 {
-		status = "okay";
-		compatible = "silicon touch,vm149C-v4l2-i2c-subdev";
-		reg = < 0x0c >;
-	};
-
-	camera1:camera-module@1 {
-		status = "okay";
-		compatible = "sony,imx219-v4l2-i2c-subdev";
-		reg = < 0x10 >;
-		device_type = "v4l2-i2c-subdev";
-		clocks = <&cru SCLK_CIFOUT>;
-		clock-names = "clk_cif_out";
-		pinctrl-names = "default", "sleep";
-		pinctrl-0 = <&cam0_default_pins>;
-		pinctrl-1 = <&cam0_sleep_pins>;
-		rockchip,pd-gpio = <&gpio2 0 GPIO_ACTIVE_LOW>;
-		rockchip,pd-gpio-is_active_low;
-		rockchip,camera-module-regulator-names = "vcc33_mipi";
-		rockchip,camera-module-regulator-voltages = <3300000>;
-		rockchip,camera-module-dovdd = "1.8v";
-		rockchip,camera-module-facing = "back";
-		rockchip,camera-module-name = "cmk-cb0695-fv1";
-		rockchip,camera-module-len-name = "lg9569a2";
-		rockchip,camera-module-fov-h = "66.0";
-		rockchip,camera-module-fov-v = "50.1";
-		rockchip,camera-module-orientation = <0>;
-		rockchip,camera-module-iq-flip = <0>;
-		rockchip,camera-module-iq-mirror = <0>;
-		rockchip,camera-module-flip = <0>;
-		rockchip,camera-module-mirror = <0>;
-
-		/* resolution.w, resolution.h, defrect.left, defrect.top, defrect.w, defrect.h */
-		rockchip,camera-module-defrect0 = <3264 2448 0 0 3264 2448>;
-		rockchip,camera-module-flash-support = <0>;
-		rockchip,camera-module-mipi-dphy-index = <0>;
-	};
-
-    m24c08@50 {
-        compatible = "at,24c08";
-        reg = <0x50>;
-    };
-};
-
-&i2c1 {
-        status = "okay";
-
-	pcm5122: pcm5122@4c {
-		#sound-dai-cells = <0>;
-		compatible = "ti,pcm5122";
-		reg = <0x4c>;
-		status = "disabled";
-	};
-};
-
-&i2c3 {
-        status = "okay";
-};
-
-&i2c4 {
-        status = "okay";
-};
-
-&spi2 {
-        status = "okay";
-        max-freq = <50000000>;
-        spidev@0 {
-                 compatible = "rockchip,spi_tinker";
-                 reg = <0>;
-                 spi-max-frequency = <50000000>;
-                 spi-cpha = <1>;
-                 //spi-cpol = <1>;
-        };
-        spidev@1 {
-                 compatible = "rockchip,spi_tinker";
-                 reg = <1>;
-                 spi-max-frequency = <50000000>;
-                 spi-cpha = <1>;
-                 //spi-cpol = <1>;
-        };
-};
-
-&i2s {
-	#sound-dai-cells = <0>;
-	status = "okay";
-};
-
 &sdio0 {
 	status = "okay";
 	clock-frequency = <50000000>;
@@ -547,10 +452,6 @@
 	supports-sdio;
 };
 
-&pwm0 {
-	status = "okay";
-};
-
 &sdmmc {
 	bus-width = <4>;
 	cap-mmc-highspeed;
@@ -735,20 +636,6 @@
 			rockchip,pins = <4 19 RK_FUNC_GPIO &pcfg_pull_none>;
 		};
 	};
-
-	cam_pins {
-		cam0_default_pins:cam0_default_pins {
-			rockchip,pins = <0 17 RK_FUNC_GPIO &pcfg_pull_none>,/*sensor power pin: GPIO0_C1 */
-			<2 0 RK_FUNC_GPIO &pcfg_pull_none>,/*sensor power down pin: GPIO2_A0*/
-			<2 11 RK_FUNC_1 &pcfg_pull_none>;/*sensor mclk: cif_clkout*/
-		};
-
-		cam0_sleep_pins:cam0_sleep_pins {
-			rockchip,pins = <0 17 RK_FUNC_GPIO &pcfg_pull_none>,/*sensor power pin: GPIO0_C1 */
-			<2 0 RK_FUNC_GPIO &pcfg_pull_none>,/* cif power down pin */
-			<2 11 RK_FUNC_GPIO &pcfg_pull_none>;/* cif_clkout */
-		};
-	};
 };
 
 &dmc {
